modules/javafx.graphics/src/main/java/javafx/scene/SpotLight.java

```diff
@@ -0,0 +1,231 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package javafx.scene;
+
+import com.sun.javafx.scene.DirtyBits;
+import com.sun.javafx.scene.NodeHelper;
+import com.sun.javafx.scene.SpotLightHelper;
+import com.sun.javafx.sg.prism.NGNode;
+import com.sun.javafx.sg.prism.NGSpotLight;
+
+import javafx.beans.property.DoubleProperty;
+import javafx.beans.property.ObjectProperty;
+import javafx.beans.property.SimpleObjectProperty;
+import javafx.geometry.Point3D;
+import javafx.scene.paint.Color;
+import javafx.scene.paint.PhongMaterial;
+
+/**
+ * A {@code PointLight} that radiates light in a cone in a specific direction. The direction of the {@code SpotLight} is
+ * defined by the {@link #directionProperty() direction} property.
+ * <p>
+ * The light cone is defined by 3 factors: an {@link #innerAngleProperty() inner angle}, an {@link #outerAngleProperty()
+ * outer angle}, and a {@link #falloffProperty() falloff} factor. For a point whose angle to the light is {@code a}, if
+ * {@code a < innerAngle} then that point receives maximum illumination, if {@code a > outerAngle} then that point
+ * receives no illumination, and if {@code innerAngle < a < outerAngle} then the illumination is determined by the
+ * formula
+ * <pre>I = pow((cos(a) - cos(outer)) / (cos(inner) - cos(outer)), falloff)</pre>
+ * which represents a drop in illumination from the inner angle to the outer angle. {@code falloff} determines the
+ * behavior of the drop.
+ * <p>
+ * <img src="doc-files/spotlight.png" alt="Image of the Spotlight">
+ * <p>
+ * While {@code innerAngle} should normally be smaller than {@code outerAngle}, it's possible to create a ring light
+ * effect by setting {@code innerAngle} to be larger than {@code outerAngle} and lowering {@code maxRange}.
+ * <p>
+ * <img src="doc-files/spotlightRing.png" alt="Image of the ring light effect">
```

Comment:
Similar to what I said about the falloff, I can remove the documentation, but I don't see a compelling reason to artificially limit the values. The equation is as described in the docs. If someone puts in values outside of physically realistic ranges they will still get whatever the equation spits out. Do you think that there would be a too-large surprise factor?

One thing to note is that the user can do the clamping themselves (or we can do it later if we don't commit in the docs), but if we clamp, the user can't unclamp.

Reply:
The main issue I have is that the specified equation matches the proposed implementation only for values that satisfy the range constraints: `0 <= innerAngle <= outerAngle <= 180`.

The reason for this is that the shader implementation compares cosines not angles, and clamps the result of the division rather than testing whether the angle is < inner or > outer. These are very reasonable implementation choices.

We have a few options that I can see:

1. List the expected range, and specify that values outside the expected range are clamped on usage.
2. List the expected range, and specify that the results are undefined for values outside the expected range.
3. Describe the equation that is actually happening.
4. Implement what the equation actually specifies, even for out-of-range values. This would constrain the implementation and likely require extra if tests in the shader.

Option 4 makes no sense to me. I don't care for 3, since it is less easy to explain and understand, and it precludes other implementations that are equivalent for in-range values but might behave differently for out-of-range values.

So that leaves option 1 or 2 as workable options. I prefer option 1, although I'll admit we aren't consistent on this point: we let the implementation do whatever it will do with out of range color values for PhongMaterial or Lights.

Reply:
I see.

I also dislike 3 and 4. I prefer 2, but if you are convinced that 1 is better I'll go with it.

Reply:
I'm OK with option 2. So that would mean removing any mention of the "ring" effect, and documenting that the valid ranges are:

```
innerAngle: [0, 180]
outerAngle: [innerAngle, 180]
```

But we wouldn't clamp them or otherwise enforce it. Instead we could add a sentence that the results are undefined (or maybe "unpredictable"?) if the angles are outside that range.